tools/codegen/dest/register_dispatch_key.py

```diff
@@ -161,9 +166,13 @@ def generate_defn(cpp_sig: CppSignature) -> str:
     globalContext().lazyInitCUDA();
     {cuda_guard_from_tensor_options}
 """
-                elif f.device_guard and device_of is not None:
+                elif f.device_guard and tensor_arg_name is not None:
```

Comment:
another approach is to still use the first `tesnor_like` parameter, and use `device_of` or `device_or_default` based on whether it's optional.